Change output behavior of buildtest build, add system log, add .err file per test

shahzebsiddiqui opened the pull request:
This PR will address the following fix
- Change output behavior #256 
- Add .err file #257 
- Add system logging #265 and remove log per test 
- Add additional logging throughout codebase
- Fix regtest issue where BuildConfig should not be called with any invalid config file, that's because all checks are done in advance in method ``discover_configs``
- Add additional system detection including environment variable, operating system details, and python path.

This first attempt to change output behavior of buildtest build.

The output looks as follows with left align all the fields.

```
gitpod /workspace/buildtest-framework $ buildtest build -c examples/script/
Config Name                    SubTest                        Status                         Config Path                   
________________________________________________________________________________________________________________________
slurm                          slurm_down_nodes_reason        PASSED                         /workspace/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     PASSED                         /workspace/buildtest-framework/examples/script/slurm.yml
zlib                           zlib_compression               PASSED                         /workspace/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /workspace/buildtest-framework/examples/script/zlib.yml
python-circle                  circle_area                    PASSED                         /workspace/buildtest-framework/examples/script/python-circle.yml
bzip2                          bzip_compression               PASSED                         /workspace/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip_version                   PASSED                         /workspace/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip_help                      PASSED                         /workspace/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip2_help                     PASSED                         /workspace/buildtest-framework/examples/script/bzip2.yml
==============================================================
                         Test summary                         
Executed 9 tests
Passed Tests: 7 Percentage: 77.77777777777777%
Failed Tests: 2 Percentage: 22.22222222222222%

```

There was some output displayed to stdout that was intended for logstream. Looking at the way ``init_logger`` was setup, it was not intended to work as it should so for now i removed a few log statement. I will cover the log issue in a separate issue to describe the issue.

The origin output as per devel head is the following

```
gitpod /workspace/buildtest-framework $ buildtest build -c examples/script/python-circle.yml 
2020-04-05 20:39:41,617 [base.py:105 -   run() ] - [DEBUG] Running setup for executor [executor-local-runlocal]
[builder-script-circle_area]
2020-04-05 20:39:41,617 [base.py:105 -   run() ] - [DEBUG] Running run for executor [executor-local-runlocal]
________________________________________________________________________________
                             start time: 2020-04-05 20:39:41.617493
                     configuration file: /workspace/buildtest-framework/examples/script/python-circle.yml
                                testdir: /workspace/buildtest-framework/.buildtest/python-circle
                               testpath: /workspace/buildtest-framework/.buildtest/python-circle/circle_area.py
                                logpath: None
________________________________________________________________________________



STAGE                                    VALUE
________________________________________________________________________________
2020-04-05 20:39:41,618 [base.py:550 - _write_test() ] - [INFO] Opening Test File for Writing: /workspace/buildtest-framework/.buildtest/python-circle/circle_area.py
[WRITING TEST]                           PASSED
[RUNNING TEST]                           PASSED
Writing results to /workspace/buildtest-framework/.buildtest/python-circle/run/circle_area_04-05-2020-20-39.out
Finished running 1 total tests.
==============================================================
                         Test summary                         
Executed 1 tests
Passed Tests: 1 Percentage: 100.0%
Failed Tests: 0 Percentage: 0.0%
```

shahzebsiddiqui committed:
change output behavior of buildtest build. First attempt at output attempt
There were some log-content print to stdout that needed to be removed.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -252,7 +252,7 @@ def __init__(self, name, recipe_config, config_file=None, testdir=None):
 
         # Start with just terminal logger (changes to file on run)
-        self._init_logger(to_file=False)
+        # self._init_logger(to_file=False)
```
@shahzebsiddiqui this will break other tests because this first logger is _just_ to the terminal, the one set later is to file. The logger needs to be present at this stage in the build. If you change one logger (to file) you just need to change the other logger (to the terminal) to do the same. Commenting this out will break things.

shahzebsiddiqui replied:
Yea i kinda realized this and wanted to get rid of the log output from stream shown in my first message in PR. I noticed the logger was passed from BuilderBase to BaseExecutor. There were few issues related to logging in #265 which i wanted to address later.

This PR was meant to address #256

shahzebsiddiqui replied:
I don't think it makes sense to have a log stream on dry run. We are doign it for time being https://github.com/shahzebsiddiqui/buildtest-framework/blob/66e49c9edaba674d06442ebfed94d9fd549c41af/buildtest/buildsystem/base.py#L528

vsoch replied:
Was this added back, so dry run has stdout logging? Just want to verify before resolving.

shahzebsiddiqui committed:
First attempt at creating a system wide buildtest logger. The file
will be written in current directory named "buildtest.log"
First attempt at capturing some logs during a typical build process.
Remove method _init_logger and show_prepare from BuilderBase. Now BuilderBase, ExecutorBase
BuilderConfig, and BuildExecutor make use of system logger.
Remove extra checks to validate configuration file in BuildConfig class
since it was already done in discover_configs method.

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -78,14 +82,6 @@ def load(self, config_file):
         """
         self.config_file = os.path.abspath(config_file)
 
-        if not os.path.exists(self.config_file):
```
Is this check still done in self._validate_global()?

shahzebsiddiqui replied:
this is not needed in ``self._validate_global()`` because it is done earlier in the ``discover_configs`` stage in particular at https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/menu/build.py#L45 you will notice that we call discover_configs before calling BuildConfig so no point in doing these checks again

https://github.com/HPC-buildtest/buildtest-framework/blob/abcfad94b69598f87f15c58e3c0f0762d9c9a99f/buildtest/menu/build.py#L87

vsoch replied:
Okay great! Just wanted to double check.

shahzebsiddiqui replied:
Okay so there is an issue with Travis see https://travis-ci.com/github/HPC-buildtest/buildtest-framework/jobs/316447547 and it is related to how regtest was written since ``discover_config`` is not used in test before invoking BuildConfig

https://github.com/shahzebsiddiqui/buildtest-framework/blob/change_output_format/tests/buildsystem/test_base.py#L22

I guess either we refactor regtest or i undo my change. What your thoughts?

vsoch replied:
The discover_config, in that it finds and validates a config file, must of course happen before you try to load it.

shahzebsiddiqui replied:
Okay so i found that culprit. If i remove the lines https://github.com/shahzebsiddiqui/buildtest-framework/blob/change_output_format/tests/buildsystem/test_base.py#L21-22 then test pass. 

I see this test would be there when no config is passed and I think what you had in mind was if user runs ``buildtest build`` it should try to search from ``~/.buildtest/site`` i.e TESTROOT, but thing is we have a check at end if no config files found that it will exit which is here https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/menu/build.py#L60

So i think it's not possible to have ``BuildConfig("")`` as a call since discover_config will catch all errors. Hope that makes sense. I ran test locally when commenting out line 21,22 and test was working. I was able to reproduce same error in Travis when the lines were uncommented.

So i'd recommend we should remove these lines.  I think there should be test for method ``discover_config`` which should catch all issues related to configuration, that way checks are done in advance before task is delegated to other classes. 
What do you think?

I think

shahzebsiddiqui replied:
Also looking at codecov https://codecov.io/gh/HPC-buildtest/buildtest-framework/src/devel/buildtest/menu/build.py we dont have any test for ``discover_config`` and i think some regtest should be written to catch all types of input for configuration (file, directory, or path relative to ``.site``). That way the test are modular per method as pose to one or two monolothic test such as one in ``buildsystem/test_base.py``

shahzebsiddiqui replied:
Okay travis is now  working with latest fix i proposed. I think this is the last remaining issue. I updated scope of PR as this will address three different issues. Let me know if this fix is acceptable and we can come to resolution.

vsoch replied:
> Okay so i found that culprit. If i remove the lines https://github.com/shahzebsiddiqui/buildtest-framework/blob/change_output_format/tests/buildsystem/test_base.py#L21-22 then test pass.

This isn't a culprit - if the user doesn't provide a valid path the test should fail with systemExit, which was working before and is exactly what that test is doing. What seems to be the issue is that discover_config is not SystemExit-ing as it should. I think the lines should be kept, and the discover_config should system exit as it was originally done.

shahzebsiddiqui replied:
Okay so would you want me to undo the change on line 81  and revert my regtest so that ``BuildTestConfig("")`` can be invoked which should raise SystemExit. 

```
if not os.path.exists(self.config_file):
```

vsoch replied:
yes, removing the check that the config file exists isn't logical. The lines that I added for both a file, and then a directory, should not be removed.

shahzebsiddiqui replied:
@vsoch i undid my changes and added test to validate ``discover_config`` See latest commit. This increases coverage of ``buildtest/menu/build.py`` which was previously at 20% now at 49%. See https://codecov.io/gh/HPC-buildtest/buildtest-framework/tree/devel/buildtest/menu

shahzebsiddiqui replied:
I think you can see the coverage in coveralls here https://coveralls.io/builds/29955036

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -274,24 +267,11 @@ def _create_test_folders(self):
         create_dir(testdir)
-        for folder in ["run", "log"]:
+        for folder in ["run"]:
             name = "%sdir" % folder
```
okay so just to clarify - this "log" folder is removed with preference for putting the buildtest.log directly in the test output folder.

shahzebsiddiqui replied:
Yes there is no log file generated in the testdir and its now just called ``buildtest.log`` in current  directory. Note for now its called ``buildtest.log`` i did this purposely so user can reference one file, it may be useful to remove the file before writing as for now its appending the content.

vsoch replied:
Perfect.

shahzebsiddiqui replied:
Just FYI. No log file in test directory now

```
(buildtest-framework) ssi29@ag-mxg-hulk090> tree .buildtest
.buildtest
├── python-circle
│   ├── circle_area.py
│   └── run
│       ├── circle_area_04-07-2020-13-19.err
│       ├── circle_area_04-07-2020-13-19.out
│       ├── circle_area_04-07-2020-13-26.err
│       ├── circle_area_04-07-2020-13-26.out
│       ├── circle_area_04-07-2020-13-27.err
│       └── circle_area_04-07-2020-13-27.out
└── slurm
    ├── run
    │   ├── slurm_down_nodes_reason_04-07-2020-13-17.err
    │   ├── slurm_down_nodes_reason_04-07-2020-13-17.out
    │   ├── slurm_down_nodes_reason_04-07-2020-14-59.err
    │   ├── slurm_down_nodes_reason_04-07-2020-14-59.out
    │   ├── slurm_down_nodes_reason_04-07-2020-15-02.err
    │   ├── slurm_down_nodes_reason_04-07-2020-15-02.out
    │   ├── slurm_down_nodes_reason_04-07-2020-15-04.err
    │   ├── slurm_down_nodes_reason_04-07-2020-15-04.out
    │   ├── slurm_not_responding_nodes_04-07-2020-13-17.err
    │   ├── slurm_not_responding_nodes_04-07-2020-13-17.out
    │   ├── slurm_not_responding_nodes_04-07-2020-14-59.err
    │   ├── slurm_not_responding_nodes_04-07-2020-14-59.out
    │   ├── slurm_not_responding_nodes_04-07-2020-15-02.err
    │   ├── slurm_not_responding_nodes_04-07-2020-15-02.out
    │   ├── slurm_not_responding_nodes_04-07-2020-15-04.err
    │   └── slurm_not_responding_nodes_04-07-2020-15-04.out
    ├── slurm_down_nodes_reason.sh
    └── slurm_not_responding_nodes.sh

4 directories, 25 files

```

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -441,9 +418,13 @@ def run_tests(self, testfile):
 
         # Change to the test directory
         os.chdir(self._get_testdir())
```
Is this does any ping to the file system to check for existence, it should be called once and then used. E.g.,
```python
testdir = self._get_testdir()
```
instead of being called multiple times to return the same folder.

shahzebsiddiqui replied:
as of now (``devel`` HEAD) its only being called twice. See https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/buildsystem/base.py 

It could be  useful to just retrieve via ``self.metadata.get("testdir")`` 

My only concern is I dont think we have os.path.expandvars set such that a test directory can be set using variable expansion.  This was done correctly in 

https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/buildsystem/base.py#L268

We should probably have test to validate something like ``buildtest build --testdir $HOME`` and see if it writes test correctly.

vsoch replied:
I think we should maybe aim to call this once, and store for the duration of the test? These sorts of setup variables we wouldn't want to weirdly change (for some buggy reason) in the middle of a test.

shahzebsiddiqui replied:
i agree and if i understand the only reason why we are ``chdir`` before running test is to ensure any generated content from the test that may include compilation, executable, or output files get stored in test directory and not in current directory which would be ugly. Unrelated to this, i noticed that we don't remove test directory before writing test. This is an issue when some test fail upon re-execution since they assume a clean directory. 

This is true for example in ``examples/script/bzip2.yml``. The point is should test directory be clean to avoid rewriting content of generated file.

vsoch replied:
I always get worried whenever scripts take charge to delete content - for example what if there was a bug with testdir and you shutil.rmtree something important? I think it makes sense to clean up the test directory on re-run, but is there some marker we can use to be absolutely sure that it's a test directory? Perhaps the output file or the folder structure? We haven't talked much about the structure of a test directory, but whatever that might encompass, this could be a good check to do.

shahzebsiddiqui replied:
The other alternate is stick a build id in the sub-directory path to avoid rewriting to same directory, that's what i did before to avoid this issue and worked nice. Previously, buildtest would purge and write again, then this was avoided by added a unique sub-directory that is incremented. Every build was then written to ``build.json`` and there was a method to retrieve total build IDs so that buildtest knows what the next build ID is. 
Adding this complete feature would require a bit more refactoring and it could be its own PR. Right now everything is working as expected.

Another alternate to an ID is use a random hex (32bit) lets say to stick as build ID, later on buildtest would need to capture build ID in some json file to retrieve test reports. Again that's out of scope for another PR and feature request.

vsoch replied:
I agree it's out of scope - I think it would be cleanest to remove the directory, but have some check that it's a build test directory, and maybe a user flag. I don't see it being useful to run multiple builds and have the folders build up (it was really messy last time) but if this approach is wanted, it would work to do a random hash each time (and make that hash very clear to the user in the screen logs). I think this is what I've seen other similar tools, either for workflows or test runs, do.

shahzebsiddiqui replied:
I'll try to open a separate issue as this was related to Stampede test i was running. Some tests couldn't rerun after first time because directory existed. I think the value of reruns would be run test for reproducible results .

One good use-case is to run acceptance test once a week, and maybe performance test once a month. Right now all the tests for stampede2 are in one directory so i can just do ``buildtest build -c tests`` and it will run everything. Or i can go by sub-directory. But eventually i'd like to have this as repeated operation so buildtest can support multiple reruns cleanly so reports can be generated based on a build ID with timestamp on when it was run. 

I think the framework should support a way to gather data from test reports which users can understand.  Running ad-hoc commands once would not be useful unless all the data is results are captured somewhere that buildtest can query and report.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -455,31 +436,35 @@ def run_tests(self, testfile):
-        # Run the test file, print output to file
-        with open(run_output_file, "w") as fd:
+        # Keep an error file
+        run_error_file = os.path.join(
```
For this, it would be more elegant to create a run_output file:

```python
run_output_file = os.path.join(	
            self.metadata.get("rundir"), self.build_id
        )
```
And then using the basename for each of .out and .err:

```python
with open(run_output_file + ".err")
   # write error

with open(run_output_file + ".out")
   # write output
```
I don't see from these changes where run_output_file and run_error_file are coming from, but that could be adopted to the above.

shahzebsiddiqui replied:
This makes sense i can refactor this stay tuned, it came to my mind also but didn't do it for first implementation.

vsoch replied:
Sounds good!

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -455,31 +436,35 @@ def run_tests(self, testfile):
-            print("{:<40} {}".format("[RUNNING TEST]", "FAILED"))
-
-        print("Writing results to " + run_output_file)
+            print(
```
Should this be FAILED if the return code isn't 0?

shahzebsiddiqui replied:
good catch that was typo on my end due to copy-paste.

vsoch commented on buildtest/executors/base.py:
```diff
@@ -101,9 +103,7 @@ def run(self, builder):
-                executor.builder.logger.debug(
-                    "Running %s for executor %s" % (step, executor)
-                )
+                # executor.builder.logger.debug("Running %s for executor %s" % (step, executor))
```
Should this be commented out?

shahzebsiddiqui replied:
Ok i fixed this locally and looked at logfile and i am getting the message 

```
2020-04-07 13:19:49,169 [base.py:106 -   run() ] - [DEBUG] Running setup for executor [executor-local-runlocal]
2020-04-07 13:19:49,169 [base.py:106 -   run() ] - [DEBUG] Running run for executor [executor-local-runlocal]

```

vsoch replied:
Cool! It might be possible to simplify that, but I'm okay with it for now :)

vsoch commented on buildtest/system.py:
```diff
@@ -23,20 +24,30 @@ def __init__(self):
-        self.system["OS_VERSION"] = distro.linux_distribution(
-            full_distribution_name=False
-        )[1]
+        self.system["os_ver"] = distro.linux_distribution(full_distribution_name=False)[
```
What is the namespace for these keys? E.g., why are we using "os_ver" which is sort of short /. choppy?

shahzebsiddiqui replied:
good point right now these are only used for logging, there was some additional work in this PR https://github.com/HPC-buildtest/buildtest-framework/pull/236/files#diff-da5c353202a52de72277768aebc926b6R35-R44 which i kept open but won't get merged. Idea was to use BuildTestSystem to detect system detail some of which is auto-detected.

As per ``os_ver`` and ``os_name`` i was trying to do a few things
1. Standardize all keys to be lower case and multi-words separated via underscore _
2. Capture OS details as part of logging, could be used for logging or reporting as part of run results.
3. Before buildtest was doing sanity check of binaries by system packages. Initially it supported rpm so i had a requirement for buildtest to only support Centos/Fedora/RHEL based system. That feature is dropped so there is no real purpose of having os_name and os_ver other than logging. The feature was called ``buildtest build --package <package-name>``. Its been dropped long ago.

vsoch replied:
Is there any reason to truncate the name? I like `os_name`, but would it make sense to do `os_version` so it's explicitly clear?

shahzebsiddiqui replied:
ill make the change to ``os_version`` makes sense. Then no truncation for time being. If you see in the other PR https://github.com/HPC-buildtest/buildtest-framework/pull/236/files#diff-da5c353202a52de72277768aebc926b6R39-R41 i can go a bit more elegant such that we can 
reference something like 

```
self.system['os']['name']
self.system['os']['version']
self.system['os']['major']
self.system['os']['major']
self.system['os']['build_number']
```

Again we can go crazy with this and would be fine with making any change that makes sense. Let me know your thoughts or a proposal

vsoch replied:
What you showed above is lovely! No worries about adding it here, as you mentioned it can be a separate PR.

shahzebsiddiqui replied:
Okay i made some changes in last commit just now to add a few extra details. Here is a snipet of the log, i think adding the environment section can be useful in troubleshooting.


```
(buildtest-framework) ssi29@ag-mxg-hulk090> head -n 40 buildtest.log
2020-04-07 15:04:54,124 [main.py:19 -  main() ] - [INFO] Starting buildtest log
2020-04-07 15:04:54,124 [system.py:28 - __init__() ] - [DEBUG] Starting System Compatibility Check
2020-04-07 15:04:54,196 [system.py:58 - init_system() ] - [INFO] Operating System: rhel
2020-04-07 15:04:54,197 [system.py:59 - init_system() ] - [INFO] Operating System Version: 7.6
2020-04-07 15:04:54,197 [system.py:60 - init_system() ] - [INFO] Operating System Major Version: 7
2020-04-07 15:04:54,197 [system.py:61 - init_system() ] - [INFO] Operating System Minor Version: 7
2020-04-07 15:04:54,197 [system.py:62 - init_system() ] - [INFO] Operating System Build Number:
2020-04-07 15:04:54,197 [system.py:64 - init_system() ] - [INFO] Session Environment Variables
2020-04-07 15:04:54,197 [system.py:65 - init_system() ] - [INFO] ________________________________________________________________________________
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] REMOTEHOST: lt4-th3-dsa.dsone.3ds.com
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] CABAQUS: /usr/abaqus60/udev/source
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] EBVERSIONZLIB: 1.2.11
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] MANPATH: /mxg-hpc/users/ssi29/easybuild/software/Python/3.7.4-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/Python/3.7.4-GCCcore-8.3.0/man:/mxg-hpc/users/ssi29/easybuild/software/libffi/3.2.1-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/XZ/5.2.4-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/SQLite/3.29.0-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/Tcl/8.6.9-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/Tcl/8.6.9-GCCcore-8.3.0/man:/mxg-hpc/users/ssi29/easybuild/software/libreadline/8.0-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/ncurses/6.1-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/zlib/1.2.11-GCCcore-8.3.0/share/man:/mxg-hpc/users/ssi29/easybuild/software/bzip2/1.0.8-GCCcore-8.3.0/man:/mxg-hpc/users/ssi29/easybuild/software/GCCcore/8.3.0/share/man:/r/cluster_software/pub_lsf/10.1/man:/usr/local/man:/usr/share/man:/usr/man:/usr/X11R6/man:/usr/openwin/man:/usr/share/man/allman:/usr/abaqus60/udev/generated/man:/usr/abaqus60/udev/object/lnx86_64/External/graphviz/man
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] FIGS: /p/marketing/figures
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] EBVERSIONPYTHON: 3.7.4
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] XDG_SESSION_ID: 144609
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] CABAQUS_O: /usr/abaqus60/udev/object/lnx86_64
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] HOSTNAME: ag-mxg-hulk090
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] EBDEVELPYTHON: /mxg-hpc/users/ssi29/easybuild/software/Python/3.7.4-GCCcore-8.3.0/easybuild/Python-3.7.4-GCCcore-8.3.0-easybuild-devel
2020-04-07 15:04:54,197 [system.py:67 - init_system() ] - [INFO] _ModuleTable003_: Zy1ocGMvdXNlcnMvc3NpMjkvZWFzeWJ1aWxkL21vZHVsZXMvYWxsL1NRTGl0ZS8zLjI5LjAtR0NDY29yZS04LjMuMC5sdWEiLFsiZnVsbE5hbWUiXT0iU1FMaXRlLzMuMjkuMC1HQ0Njb3JlLTguMy4wIixbImxvYWRPcmRlciJdPTcscHJvcFQ9e30sWyJzdGFja0RlcHRoIl09MSxbInN0YXR1cyJdPSJhY3RpdmUiLFsidXNlck5hbWUiXT0iU1FMaXRlLzMuMjkuMC1HQ0Njb3JlLTguMy4wIix9LFRjbD17WyJmbiJdPSIvbXhnLWhwYy91c2Vycy9zc2kyOS9lYXN5YnVpbGQvbW9kdWxlcy9hbGwvVGNsLzguNi45LUdDQ2NvcmUtOC4zLjAubHVhIixbImZ1bGxOYW1lIl09IlRjbC84LjYuOS1HQ0Njb3JlLTguMy4wIixbImxvYWRPcmRlciJdPTYscHJvcFQ9e30sWyJzdGFja0RlcHRo

```

I think in future it would help capture paths to launchers (srun, sbatch, mpirun, mpiexec) I had some ideas in buildtest setting to allow user to specify path to launcher paths let's say in case of mpirun you or other mpi specific launcher you dont want to rely what's in $PATH as it can change due to shell or changes to module so to ensure consistency it would be useful to let user specify a full path to launcher. Again out of scope for this task.  

For now i think it's safe to say this issue is resolved. I changed ``[os_ver]`` to ``['os']['version']``

vsoch replied:
Great! One final tweak - all of those variables should be for level DEBUG and not INFO.

shahzebsiddiqui replied:
makes sense i will make that change, should i make all of ``self.system`` be DEBUG to be consistent?

shahzebsiddiqui replied:
ill make it DEBUG for all them to be consistent.

vsoch replied:
yes.

shahzebsiddiqui committed:
remove buildtest.log before writing again, originally this was appending
the logs for repeated runs.
Change logic on .out and .err file were assigned in variable.
fix bug where buildtest was reported non-zero return code as PASSED in
output
Add log message in executor/base.py to show which executor step was run

shahzebsiddiqui committed:
capture os name, version, major/minor and build number. capture python
path and list of environment variables as part of logging

shahzebsiddiqui committed:
change system detection details to DEBUG instead of INFO in logging

shahzebsiddiqui committed:
fix regtest to address Travis build. Issue is BuildConfig should not be called with any invalid config, this check is done in discover_config method

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
undo change in BuildConfig, now BuildConfig can check
for config files to ensure regtest test/buildsystem/test_base.py is
running.
Add regtest for discover_configs.
Add extra check in discover_configs to ensure only .yaml or .yml
extension are accepted. Otherwise this causes issue later when we
need to load the YAML.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -545,7 +516,6 @@ def _write_test(self, lines=None):
         # '$HOME/.buildtest/testdir/<name>/<name>_<timestamp>.sh'
         # This will put output (latest run) in same directory - do we want this?
         testpath = os.path.expandvars(self.metadata["testpath"])
```
Should we save this with the builder so it doesn't need to be run (with expandvars) more than once? It should be done at the beginning of run/dry_run or something shared between them before that.

shahzebsiddiqui replied:
hmm, that does make sense shall i overwrite the variable ``self.metadata["testpath"]``

vsoch replied:
That would be easy enough to do - the reason to have the expansion is to honor some custom environment variable definition, but once the build has started the user can consider that "frozen" so I think updating it (and using that full path for the rest of tests) makes most sense.

shahzebsiddiqui replied:
okay just to confirm with latest change it should be working. A test for variable expansion would be something like 

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/script/slurm.yml --testdir $PWD/dummy
Config Name                    SubTest                        Status                         Config Path
________________________________________________________________________________________________________________________
slurm                          slurm_down_nodes_reason        PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
==============================================================
                         Test summary
Executed 2 tests
Passed Tests: 0 Percentage: 0.0%
Failed Tests: 2 Percentage: 100.0%
(buildtest-framework) ssi29@ag-mxg-hulk090> tree dummy/
dummy/
├── run
│   ├── slurm_down_nodes_reason_04-08-2020-21-25.err
│   ├── slurm_down_nodes_reason_04-08-2020-21-25.out
│   ├── slurm_not_responding_nodes_04-08-2020-21-25.err
│   └── slurm_not_responding_nodes_04-08-2020-21-25.out
├── slurm_down_nodes_reason.sh
└── slurm_not_responding_nodes.sh

1 directory, 6 files

```

vsoch replied:
Why does it say there are 2 percentage failed tests (then 100%) but report them above as PASSED?

vsoch replied:
For the first part I understand it’s a count followed by “Percentage 100%” so it’s just a confusing read for the user. Can we do:

2/2: 100% 

Or something simple like that!

shahzebsiddiqui replied:
Let me look into this, good catch i didn't pay attention hold on

shahzebsiddiqui replied:
okay this is fixed now i had typo again for some reason i was reporting test as PASSED

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/script/slurm.yml
Config Name                    SubTest                        Status                         Config Path
________________________________________________________________________________________________________________________
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
==============================================================
                         Test summary
Executed 2 tests
Passed Tests: 0/2 Percentage: 0.0%
Failed Tests: 2/2 Percentage: 100.0%

(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/script/python-circle.yml
Config Name                    SubTest                        Status                         Config Path
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
==============================================================
                         Test summary
Executed 1 tests
Passed Tests: 1/1 Percentage: 100.0%
Failed Tests: 0/1 Percentage: 0.0%

```

vsoch replied:
Looks good!

vsoch commented on buildtest/system.py:
```diff
@@ -23,20 +24,55 @@ def __init__(self):
+        self.system["os"]["minor_version"] = distro.major_version(best=True)
+        self.system["os"]["build_number"] = distro.build_number(best=True)
+        self.system["env"] = dict(os.environ)
+        self.system["python"] = shutil.which("python") or None
```
I think if shutil.which() can't find something it already returns None.

```python
import shutil                                                           
res = shutil.which('nope')                                              
res == None                                                             
True
```
so the "or None" is not necessary.

vsoch replied:
Also now we need to be very careful for what is done with:
```python
self.system["env"] = dict(os.environ)
```
because what goes in the environment? Secrets, of course.

shahzebsiddiqui replied:
i fixed the shutil issue. 

Oh yea i know if i figure capturing environment shell can be useful for debugging for both end-users and maintainers

vsoch commented on buildtest/system.py:
```diff
@@ -23,20 +24,55 @@ def __init__(self):
+        self.logger.info(f"Operating System: {self.system['os']['name']}")
+        self.logger.info(f"Operating System Version: {self.system['os']['version']}")
+        self.logger.info(
+            f"Operating System Major Version: {self.system['os']['major_version']}"
```
Is there any reason to not assemble a complete version string and have a single, clean message for OS version over printing a line for each?

shahzebsiddiqui replied:
no particular reason, just providing extra detail. I mean this is sufficient also at the moment it not providing any extra value other that additional detail. 

```
 distro.linux_distribution()
('Red Hat Enterprise Linux Server', '7.6', 'Maipo')
```

vsoch replied:
Yes let’s do that, it’s more human readable as a single line. If you want to join with a space to avoid the extra layer of data structure that works too.

shahzebsiddiqui replied:
okay this is fixed now it looks like this now

```
(buildtest-framework) ssi29@ag-mxg-hulk090> cat buildtest.log | grep Operating
2020-04-08 22:00:23,242 [system.py:49 - init_system() ] - [DEBUG] Operating System: Red Hat Enterprise Linux Server 7.6 Maipo

```

vsoch replied:
Much improved.

vsoch commented on buildtest/system.py:
```diff
@@ -23,20 +24,55 @@ def __init__(self):
+
+        self.logger.debug("Session Environment Variables")
+        self.logger.debug("{:_<80}".format(""))
+        for k, v in self.system["env"].items():
```
This is actually a security risk - we should not be printing environment variables to a log file. If absolutely necessary, just print the keys.

vsoch replied:
I would suggest namespacing here - for buildtest envars, we prefix with BUILDTEST and then we at least know the ones that are safe to skip / print, and we can filter os.environ down to anything that starts with `BUILDTEST` or another white listed set.

shahzebsiddiqui replied:
Yea i used to have buildtest variables prefixed with BUILDTEST that's not the case, if you want i can just print the keys without the value.

vsoch replied:
Ok let’s open an issue for the environment namespace, other than filtering we don’t want a conflict of envars with some other tool that happens to use a variable with the same name.

shahzebsiddiqui replied:
okay for now i print the keys and its fixed in the logs.

Here is a snippet

```
2020-04-08 21:25:52,375 [system.py:70 - init_system() ] - [DEBUG] Session Environment Variables
2020-04-08 21:25:52,375 [system.py:71 - init_system() ] - [DEBUG] ________________________________________________________________________________
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] REMOTEHOST
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] CABAQUS
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] EBVERSIONZLIB
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] MANPATH
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] FIGS
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] EBVERSIONPYTHON
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] XDG_SESSION_ID
2020-04-08 21:25:52,375 [system.py:73 - init_system() ] - [DEBUG] CABAQUS_O

```

shahzebsiddiqui committed:
logging only environment name but not value due to secret data
fix os.path.expandvars so that it is called once when
self.metadata["testpath"] is set
fix minor bugs

shahzebsiddiqui committed:
add coverage to file buildtest/buildsystem/schemas/utils.py
Ideally we should have test one per module file if possible. Coverage
reported a decrease in this file. Now coverage is increased from 75% --> 80%

shahzebsiddiqui committed:
remove logging of all operating system name, just have one entry now.
Before we printed major, minor, build number, version, etc...
Typo in build output, was reported test PASSED with non-zero return
code.
Fixed output of Test Summary

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -441,45 +426,49 @@ def run_tests(self, testfile):
 
-        # Run the test file, print output to file
-        with open(run_output_file, "w") as fd:
+        # Keep an error file
```
Having run_output_file and run_error_file is redundant, you can remove the second and then add an extension for err or out to the own variable.

shahzebsiddiqui replied:
thats quite odd, i recall making this change but for some reason this was not made something must have happened. Let me fix this quick

shahzebsiddiqui replied:
This is fixed now

vsoch commented on tests/buildsystem/schemas/test_utils.py:
```diff
@@ -0,0 +1,22 @@
+def test_load_schema_invalid_ext():
+
+    # invalid file extension should fail
+    load_schema(os.path.join(root, "README.rst"))
```
One thing we will need to think about for another PR is how to handle skipping over yaml files that are discovers that aren’t intended for buildtest.

shahzebsiddiqui replied:
indeed good discussion for another PR :)

shahzebsiddiqui committed:
remove unused variable

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch reviewed (approved):
Great work @shahzebsiddiqui ! I think this is ready for merge :)
